Local storage

peterbertel opened the pull request:
The pull request introduces the initial implementation of the local storage. Each new Day will be saved to a backend database json file. The contents of this file will persist until the app is deleted from the user's device. The new BackendService invokes each of these actions for submitting data to and retrieving data from the database.

This pull request will also update the new_summary.dart page, which will create new Day objects based on the contents in the backend database file.

For testing purposes, add several new days by refreshing the app after each successful entry. In a future update, we will update the logic to only allow one entry per day.

peterbertel committed:
Initial commit of the backend service

peterbertel committed:
Merge pull request #22 from peterbertel/local-storage

Initial commit of the backend service

peterbertel committed:
Fix merge conflicts

peterbertel committed:
Merge branch 'master' into local-storage

peterbertel committed:
Add hive dependencies

peterbertel committed:
Remove localstorage package usage in backend_service.dart

peterbertel committed:
Remove hive dependencies

peterbertel committed:
Add sembast dependency

peterbertel committed:
Add the backend service to main.dart and update the providers to use 'create' syntax rather than 'builder'

peterbertel committed:
Remove backend service from activities.dart and add initial backend service functions for reading and writing data to/from a local file

peterbertel committed:
Save new Day entries as JSON strings to the backend database file

This commit makes the following changes: update the DayModel toJson() function so the returned data is well-formatted for the jsonEncode() method in the BackendService can properly convert the json into a string; update questions_controller.dart to initialize a new BackendService object, add a customized timestamp for the day's id (month, day, year, miliseconds), and call the addDay() method when the user navigates to the summary page from the final question; update the BackendService to add the new day data to the existing contents from the backend database, rather than overwritting the data

peterbertel committed:
Update questions_controller.dart so the day's date and id are only set once before adding that day to the backend database

peterbertel committed:
Convert new_summary day widget to be a ListView builder

Build the child widgets for each day by passing in an array of all the days into a ListView builder instead of explicitly defining each child widget.

peterbertel committed:
Remove sembast dependency

peterbertel committed:
Add dependency for loading widget

peterbertel committed:
Add progress_indicator to the list of dependencies

peterbertel committed:
Update Day.fromMap method to properly set the date and feelings/activities sets from the snapshot

peterbertel committed:
Update getDays method in the backend_service to return a list of Days created from the database file

peterbertel committed:
Use a futurebuilder to display the Day objects retrieved from the backend database

Also display a loading progress indicator until the results appear. The Days retrieved from the database will be shown in reverse order to show the newest Day entry first

peterbertel committed:
Resolve race condition involving reading new Days from the database immediately after writing a new one

Change the addDay method in backend_service.dart to return a File object. Modify questions_controller.dart to only navigate to the new_summary.dart page when the File is returned from the addDay method. Delete the temporary solution in new_summary.dart that previously added the missing Day if it did not exist in the list of Days returned from the database.

peterbertel committed:
Cleanup backend_service.dart before pull request

peterbertel committed:
Merge branch 'master' into local-storage

ewynx commented on lib/pages/questions/questions_controller.dart:
```diff
@@ -35,24 +38,25 @@ class _QuestionsControllerState extends State<QuestionsController> {
-        print(day.date.toLocal());
-        print(day.date.toUtc().toLocal());
-        await dayService.addDocument(day.toJson());
+        // await dayService.addDocument(day.toJson());
```
what should happen with this line?

peterbertel replied:
It wasn't providing any extra functionality for the backend storage, so I had originally just commented it out. I think we can safely delete it, so I did in [this commit](https://github.com/ewynx/shelter_in_place/pull/24/commits/53ca9c64c536a774cb80555dbeaf4d3205193c59).

ewynx commented on lib/pages/summary/new_summary.dart:
```diff
@@ -28,20 +31,22 @@ class NewSummary extends StatelessWidget {
-        activities: acts.take(5).toList().toSet(),
-        note:
-            'Today was a pretty good day! I read a book and went for a run by myself.');
+    // Day day1 = Day(
```
can be removed?

ewynx replied:
actually the definition of all days day1, day2 etc can be removed!

peterbertel replied:
I deleted all the unused Day objects in [this commit](https://github.com/ewynx/shelter_in_place/pull/24/commits/53ca9c64c536a774cb80555dbeaf4d3205193c59).

ewynx commented on lib/pages/summary/new_summary.dart:
```diff
@@ -126,6 +131,10 @@ class NewSummary extends StatelessWidget {
         activities: acts.take(2).toList().toSet(),
         note: 'This is the second day');
 
+    // List<Day> days = [day1, day2, day3, day4, day5, day6, day7, day8, day9, day10, day11];
```
dito :D

peterbertel replied:
Deleted in [this commit](https://github.com/ewynx/shelter_in_place/pull/24/commits/53ca9c64c536a774cb80555dbeaf4d3205193c59).

ewynx commented on lib/services/backend_service.dart:
```diff
@@ -0,0 +1,61 @@
+
+  Future<String> getDatabasePath() async {
+    final directory = await getApplicationDocumentsDirectory();
+    // final directory = await getTemporaryDirectory();
```
can this be removed?

peterbertel replied:
Yep, I removed it in [this commit](https://github.com/ewynx/shelter_in_place/pull/24/commits/53ca9c64c536a774cb80555dbeaf4d3205193c59).

ewynx reviewed (changes requested):
Works like a charm!! Awesome 👍 
I added some comments, it's probably code that can all be removed, so just for the sake of cleanness :)

peterbertel committed:
Remove unnecessary code in preparation for merging in local-storage changes

Remove commented out code related to backend storage and delete unused Day objects on the new_summary.dart page

peterbertel commented:
@ewynx I deleted all the unnecessary code you referenced in your comments, so I am going to merge in the changes.
